Mathlib/Order/SuccPred/Basic.lean

```diff
@@ -0,0 +1,1498 @@
+
+variable [OrderTop α]
+
+@[simp]
+theorem succ_top : succ (⊤ : α) = ⊤ := by
+  rw [succ_eq_iff_isMax, isMax_iff_eq_top]
+#align order.succ_top Order.succ_top
+
+--Porting note: removing @[simp],`simp` can prove it
+theorem succ_le_iff_eq_top : succ a ≤ a ↔ a = ⊤ :=
+  succ_le_iff_isMax.trans isMax_iff_eq_top
+#align order.succ_le_iff_eq_top Order.succ_le_iff_eq_top
+
+--Porting note: removing @[simp],`simp` can prove it
+theorem lt_succ_iff_ne_top : a < succ a ↔ a ≠ ⊤ :=
+  lt_succ_iff_not_isMax.trans not_isMax_iff_ne_top
+#align order.lt_succ_iff_ne_top Order.lt_succ_iff_ne_top
+
+end OrderTop
+
+section OrderBot
+
+variable [OrderBot α]
+
+--Porting note: removing @[simp],`simp` can prove it
+theorem lt_succ_bot_iff [NoMaxOrder α] : a < succ ⊥ ↔ a = ⊥ := by rw [lt_succ_iff, le_bot_iff]
+#align order.lt_succ_bot_iff Order.lt_succ_bot_iff
+
+theorem le_succ_bot_iff : a ≤ succ ⊥ ↔ a = ⊥ ∨ a = succ ⊥ := by
+  rw [le_succ_iff_eq_or_le, le_bot_iff, or_comm]
+#align order.le_succ_bot_iff Order.le_succ_bot_iff
+
+variable [Nontrivial α]
+
+theorem bot_lt_succ (a : α) : ⊥ < succ a :=
+  (lt_succ_of_not_is_max not_isMax_bot).trans_le <| succ_mono bot_le
+#align order.bot_lt_succ Order.bot_lt_succ
+
+theorem succ_ne_bot (a : α) : succ a ≠ ⊥ :=
+  (bot_lt_succ a).ne'
+#align order.succ_ne_bot Order.succ_ne_bot
+
+end OrderBot
+
+end PartialOrder
+
+/-- There is at most one way to define the successors in a `PartialOrder`. -/
+instance [PartialOrder α] : Subsingleton (SuccOrder α) :=
+  ⟨by
+    intro h₀ h₁
+    ext a
+    by_cases ha : IsMax a
+    · exact (@IsMax.succ_eq _ _ h₀ _ ha).trans ha.succ_eq.symm
+    · exact @Covby.succ_eq _ _ h₀ _ _ (covby_succ_of_not_is_max ha)⟩
+
+section CompleteLattice
+
+variable [CompleteLattice α] [SuccOrder α]
+
+theorem succ_eq_infi (a : α) : succ a = ⨅ (b) (_h : a < b), b := by
+  refine' le_antisymm (le_infᵢ fun b => le_infᵢ succ_le_of_lt) _
+  obtain rfl | ha := eq_or_ne a ⊤
+  · rw [succ_top]
+    exact le_top
+  exact infᵢ₂_le _ (lt_succ_iff_ne_top.2 ha)
+#align order.succ_eq_infi Order.succ_eq_infi
+
+end CompleteLattice
+
+/-! ### Predecessor order -/
+
+
+section Preorder
+
+variable [Preorder α] [PredOrder α] {a b : α}
+
+/-- The predecessor of an element. If `a` is not minimal, then `pred a` is the greatest element less
+than `a`. If `a` is minimal, then `pred a = a`. -/
+def pred : α → α :=
+  PredOrder.pred
+#align order.pred Order.pred
+
+theorem pred_le : ∀ a : α, pred a ≤ a :=
+  PredOrder.pred_le
+#align order.pred_le Order.pred_le
+
+theorem min_of_le_pred {a : α} : a ≤ pred a → IsMin a :=
+  PredOrder.min_of_le_pred
+#align order.min_of_le_pred Order.min_of_le_pred
+
+theorem le_pred_of_lt {a b : α} : a < b → a ≤ pred b :=
+  PredOrder.le_pred_of_lt
+#align order.le_pred_of_lt Order.le_pred_of_lt
+
+theorem le_of_pred_lt {a b : α} : pred a < b → a ≤ b :=
+  PredOrder.le_of_pred_lt
+#align order.le_of_pred_lt Order.le_of_pred_lt
+
+@[simp]
+theorem le_pred_iff_is_min : a ≤ pred a ↔ IsMin a :=
```

Comment:
Please review the entire file for `is_max`/`is_min`

Reply:
done!